src/cleanfigure.m

```diff
@@ -296,36 +296,93 @@ function pruneOutsideBox(meta, handle)
     % in p and determine whether the line between the pair crosses the box.
     n = size(data, 1);
     out = false(n-1, 1);
+    % Only check if there is more than 1 point    
+    if n>1
+        % One of the neighbors is inside the box and the other is finite
+        nextVisible = (dataIsInBox(2:end)   & all(isfinite(data(1:end-1)),2));
+        thisVisible = (dataIsInBox(1:end-1) & all(isfinite(data(2:end)),2));
 
-    [bottomLeft, topLeft, bottomRight, topRight] = corners(xLim, yLim);
+        % Get the corner coordinates
+        [bottomLeft, topLeft, bottomRight, topRight] = corners(xLim, yLim);
 
-    for k = 1:n-1
-        this = data(k  , :);
-        next = data(k+1, :);
+        % Calculate the difference between two consequtive points
+        dataDiff = diff(data);
 
-        % One of the neighbors is inside the box and the other is finite
-        nextVisible = (dataIsInBox(k+1) && all(isfinite(this)));
-        thisVisible = (dataIsInBox(k)   && all(isfinite(next)));
-
-        % Check whether the line connecting this point and the next one
-        % intersects with any of the borders of the drawn axis
-        left   = segmentsIntersect(this, next, bottomLeft , topLeft);
-        right  = segmentsIntersect(this, next, bottomRight, topRight);
-        bottom = segmentsIntersect(this, next, bottomLeft , bottomRight);
-        top    = segmentsIntersect(this, next, topLeft    , topRight);
-
-        % The segment is visible when any of the following hold:
-        %  - this point is visible in the axis and the next is finite
-        %  - this point is finite and the next is visible in the axis
-        %  - the segment connecting this and the next point crosses a border
-        out(k) = thisVisible || nextVisible || left || right || top || bottom;
+        left   = segmentsIntersect(data, dataDiff, bottomLeft , topLeft);
+        right  = segmentsIntersect(data, dataDiff, bottomRight, topRight);
+        bottom = segmentsIntersect(data, dataDiff, bottomLeft , bottomRight);
+        top    = segmentsIntersect(data, dataDiff, topLeft    , topRight);
+
+        % Check the result
+        out = any([thisVisible, nextVisible, left, right, top, bottom],2);
     end
 end
 % =========================================================================
-function out = segmentsIntersect(X1, X2, X3, X4)
+function out = segmentsIntersect(data, dataDiff, X3, X4)
   % Checks whether the segments X1--X2 and X3--X4 intersect.
-  lambda = crossLines(X1, X2, X3, X4);
-  out = all(lambda > 0.0) && all(lambda < 1.0);
+  % Given four points X_k=(x_k,y_k), k\in{1,2,3,4}, and the two lines defined
```

Comment:
You should adapt the comment here as well.
Have you checked that this provides a speedup?
Because passing the data and derivative, might be computationally advantageous, but it seems to introduce quite a bit of extra calculations (and hence complexity and possibility of bugs) here.

Before, the function was really simple and used functionality that was already present in `crossLines` instead of trying to recompute everything in a different way.